editor/src/messages/portfolio/document/utility_types/network_interface.rs

```diff
@@ -464,6 +500,104 @@ impl NodeNetworkInterface {
+	/// 2. If the node is a network:
+	///    - Iterates through its exports (skipping the first/primary export).
+	///    - For each export, traverses the network until reaching a protonode or terminal condition.
+	///    - Determines the output type based on the final node/value encountered.
+	/// 3. Collects and returns all resolved types.
+	///
+	/// # Note
+	///
+	/// This function assumes that export indices and node IDs always exist within their respective
+	/// collections. It will panic if these assumptions are violated.
+	pub fn output_types(&self, node_id: &NodeId, network_path: &[NodeId]) -> Vec<Option<Type>> {
+		let Some(network) = self.network(network_path) else {
+			log::error!("Could not get network in output_types");
+			return Vec::new();
+		};
+		let Some(node) = network.nodes.get(node_id) else {
+			log::error!("Could not get node {node_id} in output_types");
+			return Vec::new();
+		};
+
+		let mut output_types = Vec::new();
+
+		// If the node is not a protonode, get types by traversing across exports until a proto node is reached.
+		if let graph_craft::document::DocumentNodeImplementation::Network(internal_network) = &node.implementation {
+			for export in internal_network.exports.iter() {
+				match export {
+					NodeInput::Node {
+						node_id: nested_node_id,
+						output_index,
+						..
+					} => {
+						let nested_output_types = self.output_types(nested_node_id, &[network_path, &[*node_id]].concat());
+						let Some(nested_nodes_output_types) = nested_output_types.get(*output_index) else {
+							log::error!("Could not get nested nodes output in output_types");
+							return Vec::new();
+						};
+						output_types.push(nested_nodes_output_types.clone());
+					}
+					NodeInput::Value { tagged_value, .. } => {
+						output_types.push(Some(tagged_value.ty()));
+					}
+
+					NodeInput::Network { .. } => {
+						// https://github.com/GraphiteEditor/Graphite/issues/1762
+						log::error!("Network input type cannot be connected to export");
+						return Vec::new();
+					}
+					NodeInput::Scope(_) => todo!(),
+					NodeInput::Inline(_) => todo!(),
+				}
+			}
+		} else if let graph_craft::document::DocumentNodeImplementation::ProtoNode(protonode) = &node.implementation {
+			let node_id_path = &[network_path, &[*node_id]].concat();
+			let primary_output_type = self.resolved_types.types.get(node_id_path).map(|ty| ty.output.clone()).or_else(|| {
+				let Some(node_io_hashmap) = NODE_REGISTRY.get(protonode) else {
+					log::error!("Could not get hashmap for proto node: {protonode:?}");
+					return None;
+				};
+
+				let mut all_node_io_types = node_io_hashmap.keys().collect::<Vec<_>>();
+				all_node_io_types.sort_by_key(|node_io_types| {
+					let mut hasher = DefaultHasher::new();
+					node_io_types.hash(&mut hasher);
+					hasher.finish()
+				});
+				let Some(node_types) = all_node_io_types.first() else {
+					log::error!("Could not get node_types from hashmap");
+					return None;
+				};
```

Comment:
instead of doing a sort and then taking the first value, you could just use a min_by_key here